
<launch>
	<!-- 从机机器人序号，slave1代表从机1号，slave2代表从机2号，以此类推-->
	<arg name="slave_robot" value="slave1"/> 

	<!-- 在主机（领航者）的运动轨迹中分离出从机的期望跟随坐标  -->
	<arg name="slave_x" value="-0.8"/> 		<!-- slave_x为从机（跟随者）的期望x坐标，主机前方为正方向，单位：m  -->
	<arg name="slave_y" value="0.8"/> 		<!-- slave_y为从机（跟随者）的期望y坐标，主机左方为正方向，单位：m  -->
	<node pkg="tf" type="static_transform_publisher" name="base_to_$(arg slave_robot)" args="$(arg slave_x) $(arg slave_y) 0 0 0 0   base_footprint $(arg slave_robot) 100" />

	<!-- 是否开启避障功能 默认开启-->
	<arg name="avoidance" value="true"/>

	<!-- 定义命令空间，区分不同的从机 -->
	<group ns ="$(arg slave_robot)">
		<param name="max_vel_x" value="0.8" /> 		<!-- 最大线速度限制（绝对值） -->  
		<param name="min_vel_x" value="0.05" /> 	<!-- 最小线速度限制（绝对值） -->
		<param name="max_vel_theta" value="0.8" /> 	<!-- 最大角速度限制（绝对值） -->  
		<param name="min_vel_theta" value="0.05" /> <!-- 最小角速度限制（绝对值） -->  
		<!-- 用于连通TF，left_link等...  -->
	    <param name = "tf_prefix" value="$(arg slave_robot)"/>
	    
		<!-- 开启机器人底层相关节点  -->
		<include file="$(find tringai_multi)/launch/turn_on_tringai_robot.launch" >
		  <arg name="slave_robot" value="$(arg slave_robot)"/>
		  <arg name="slave_x" value="$(arg slave_x)"/> 		<!-- 初始位置 x坐标 -->
		  <arg name="slave_y" value="$(arg slave_y)"/> 		<!-- 初始位置 y坐标 -->
		</include>

		<!-- 开启避障 -->
		<group if="$(arg avoidance)"> 
			<node name='laser_tracker' pkg="simple_follower" type="laserTracker.py"> <!-- 提取最近的障碍物距离信息 -->
			  <param name="winSize" value="2" type="int" />
			  <param name="deltaDist" value="0.2" type="double" />
			</node>
			<node pkg="tringai_multi" type="multi_avoidance" name="multi_avoidance" output="screen" > <!-- 发布速度信息 -->
				<param name="avoidance_kv" type="double" value="0.4" /> 		<!-- 修正从车线速度的参数，参数越大，对从车减速作用越强 -->
				<param name="avoidance_kw" type="double" value="0.4" /> 		<!-- 修正从车角速度的参数，参数越大，调节从车角度，使车头偏离障碍物的作用越强 -->
				<param name="safe_distence" type="double" value="0.6" /> 		<!-- 安全距离界限 -->
				<param name="danger_distence" type="double" value="0.3" /> 		<!-- 危险距离界限 -->
			</node>
		</group>

	    <!-- 开启用于导航的自适应蒙特卡洛定位amcl-->
	    <include file="$(find tringai_multi)/launch/include/amcl.launch" >
	      <arg name="use_map_topic" value="true"/>
	      <arg name="slave_robot" value="$(arg slave_robot)"/>
	    </include>

	      <!-- 从机1号的位置监听与速度控制节点  -->
	    <node pkg="tringai_multi" type="slave_tf_listener" name="slave_tf_listener" output="screen" > <!-- 命名空间slave1代表从机1号 -->
	      <!-- 参数调节  -->
		  <param name="k_v" type="double" value="1.5" /> 			<!-- 调节前后方向偏差时，k_v越大，线速度越大 -->  
		  <param name="k_l" type="double" value="1.0" /> 			<!-- 调节左右方向偏差时，k_l越大，角速度越大 -->  
		  <param name="k_a" type="double" value="1.0" />			<!-- 调节角度偏差时，k_a越大，角速度越大 -->  
		  <param name="slave_x" type="double" value="$(arg slave_x)" />
		  <param name="slave_y" type="double" value="$(arg slave_y)" />
	      <param name="base_to_slave" type="string" value="/$(arg slave_robot)" /> <!-- 期望跟随坐标的TF坐标名 -->      
	      <param name="base_frame" type="string" value="base_footprint" /> <!-- 从机的TF坐标名 --> 
	      <remap from="cmd_vel_ori" to="cmd_vel" if="$(eval avoidance == false)"/>
	    </node> 

	</group>
	
</launch>
